### PR TITLE
Restart pds application pod during resource update

### PR DESCRIPTION
**Pull request by gauraa2**
This PR has the test RestartAppDuringResourceUpdate

Deploy data service
Update the resource config, while upgrade is happening delete few pods.
Validate post deletion all the pods come up successfully.

Logs: https://aetos.pwx.purestorage.com/resultSet/testSetID/164571

### CODE REVIEW

File: tests/pds/resiliency_test.go [hunk 1]
```diff
@@ -94,3 +94,57 @@ var _ = Describe("{KillDeploymentControllerDuringDeployment}", func() {
...
+	JustBeforeEach(func() {
+		StartTorpedoTest("RestartAppDuringResourceUpdate", "Restart application pod during resource update", pdsLabels, 0)
+		pdslib.MarkResiliencyTC(true, false)
+	})
+
+	It("Deploy Dataservices", func() {
+		var deployments = make(map[PDSDataService]*pds.ModelsDeployment)
+		Step("Deploy Data Services", func() {
+			for _, ds := range params.DataServiceToTest {
+				if ds.Name != postgresql {
```
Comment (Madanagopal19) [inline tests/pds/resiliency_test.go:108]: remove this check, we need to run the test for all data services

File: tests/pds/resiliency_test.go [hunk 2]
```diff
@@ -94,3 +94,57 @@ var _ = Describe("{KillDeploymentControllerDuringDeployment}", func() {
...
+					continue
+				}
+				Step("Deploy and validate data service", func() {
+					isDeploymentsDeleted = false
```
Comment (Madanagopal19) [inline tests/pds/resiliency_test.go:112]: this flag is not required

File: tests/pds/resiliency_test.go [hunk 3]
```diff
@@ -94,3 +94,57 @@ var _ = Describe("{KillDeploymentControllerDuringDeployment}", func() {
...
+
+				err = pdslib.InduceFailureAfterWaitingForCondition(deployment, namespace, 0)
+				log.FailOnError(err, fmt.Sprintf("Error while pod restart during Resource update %v", *deployment.ClusterResourceName))
+
+				err = pdslib.ValidateDataServiceDeployment(deployment, namespace)
+				log.FailOnError(err, "error on ValidateDataServiceDeployment")
+			}
+		})
+	})
+})
```
Comment (Madanagopal19) [inline tests/pds/resiliency_test.go:150]: Add aftereach block and end torpedo test also close the resiliency channel

File: drivers/pds/lib/utils.go [hunk 4]
```diff
@@ -2586,3 +2586,30 @@ func GetPodsOfSsByNode(SSName string, nodeName string, namespace string) ([]core
...
+	for _, template := range resourceTemplates {
+		log.Infof("template - %v", template.GetName())
+		if strings.ToLower(template.GetName()) == strings.ToLower(resourceTemplate) {
+			resourceTemplateId = template.GetId()
+		}
+	}
+	if resourceTemplateId == "" {
+		log.FailOnError(fmt.Errorf("resource template - {%v} , not found", resourceTemplate), "Error while getting the required resource setting template.")
+	}
+	log.Infof("Deployment details.")
```
Comment (Madanagopal19) [inline drivers/pds/lib/utils.go:2605]: combine both the loggers

File: drivers/pds/lib/resiliency_utils.go [hunk 5]
```diff
@@ -111,6 +114,16 @@ func InduceFailureAfterWaitingForCondition(deployment *pds.ModelsDeployment, nam
 			InduceFailure(FailureType.Type, namespace)
 		}
 		ExecuteInParallel(func1, func2)
+	case RestartAppDuringResourceUpdate:
+		log.InfoD("Entering to check if Data service has %v active pods. "+
+			"Once it does, we restart application pods", checkTillReplica)
+		func1 := func() {
+			UpdateDeploymentResourceConfig(deployment, "medium")
```
Comment (Madanagopal19) [inline drivers/pds/lib/resiliency_utils.go:121]: take template name as a constant

File: drivers/pds/lib/utils.go [hunk 6]
```diff
@@ -2586,3 +2586,30 @@ func GetPodsOfSsByNode(SSName string, nodeName string, namespace string) ([]core
...
 }
+
+func UpdateDeploymentResourceConfig(deployment *pds.ModelsDeployment, resourceTemplate string) error {
+	var resourceTemplateId string
+	resourceTemplates, err := components.ResourceSettingsTemplate.ListTemplates(*deployment.TenantId)
+	if err != nil {
+		return err
+	}
+	for _, template := range resourceTemplates {
+		log.Infof("template - %v", template.GetName())
```
Comment (Madanagopal19) [inline drivers/pds/lib/utils.go:2597]: use log.Debugf incase if you are printing the names for debugging purpose

**Review by Madanagopal19 — CHANGES_REQUESTED** [review]

File: drivers/pds/lib/utils.go [hunk 7]
```diff
@@ -2586,3 +2586,30 @@ func GetPodsOfSsByNode(SSName string, nodeName string, namespace string) ([]core
...
+	if resourceTemplateId == "" {
+		log.FailOnError(fmt.Errorf("resource template - {%v} , not found", resourceTemplate), "Error while getting the required resource setting template.")
+	}
+	log.Infof("Deployment details.")
+	log.Infof("Ds id- %v, appConfigTemplateID - %v, imageId - %v, Node count -%v, resourceTemplateId- %v ", deployment.GetId(),
+		appConfigTemplateID, deployment.GetImageId(), deployment.GetNodeCount(), resourceTemplateId)
+	_, err = components.DataServiceDeployment.UpdateDeployment(deployment.GetId(),
+		appConfigTemplateID, deployment.GetImageId(), deployment.GetNodeCount(), resourceTemplateId, nil)
+	if err != nil {
+		return err
```
Comment (dbhatnagar-px) [inline drivers/pds/lib/utils.go:2611]: Add this error in CapturedErrors Channel

File: drivers/pds/lib/resiliency_utils.go [hunk 8]
```diff
@@ -221,3 +234,25 @@ func KillDeploymentPodDuringDeployment(ns string) error {
 	}
 	return testError
 }
+
+func RestartApplicationDuringResourceUpdate(ns string) error {
+	var ss *v1.StatefulSet
+	ss, testError := k8sApps.GetStatefulSet(deployment.GetClusterResourceName(), ns)
+	if testError != nil {
+		CapturedErrors <- testError
```
Comment (dbhatnagar-px) [inline drivers/pds/lib/resiliency_utils.go:242]: Return testError as well

**Review by dbhatnagar-px — CHANGES_REQUESTED** [review]

File: drivers/pds/lib/utils.go [hunk 9]
```diff
@@ -2586,3 +2586,30 @@ func GetPodsOfSsByNode(SSName string, nodeName string, namespace string) ([]core
...
+	}
+	ResiliencyCondition <- true
```
Comment (dbhatnagar-px) [inline drivers/pds/lib/utils.go:2613]: ResiliencyCondition should be set only after validating that Update has started for atleast 1 pod

**Review by dbhatnagar-px — CHANGES_REQUESTED** [review]

File: tests/pds/resiliency_test.go (hunk 2 above)
Comment (Madanagopal19) [reply tests/pds/resiliency_test.go:310]: This is not addressed

**Review by Madanagopal19 — CHANGES_REQUESTED** [review]

**Comment by Madanagopal19** [comment]
Add test execution logs in description

**Review by dbhatnagar-px — APPROVED** [review]
Lgtm

**Review by Madanagopal19 — APPROVED** [review]
lgtm

**Review by aravindranath14 — APPROVED** [review]
Looks Good

**Pull request merged**